### PR TITLE
Windows documentation overhaul

### PR DESCRIPTION
**Pull request by marosset**
This PR moves all content out of /docs/setup/production-environment/windows to new locations. 
Where possible we merged Windows information into the existing docs.

This is a big PR but most of the contents have already been reviewed
- https://github.com/kubernetes/website/pull/31900 for storage related docs
- https://github.com/kubernetes/website/pull/31852 for networking related docs
- https://github.com/kubernetes/website/pull/31899 for troubleshooting section

I was initially hoping to get all of these changes in with the 1.24 release branch but in order to ease pressure on the release team we all decided to merge these changes after the 1.24 release.

Part of https://github.com/kubernetes/website/issues/31428

/label refactor
/sig windows
/assign @jsturtevant @sftim 
/cc @tengqm @aravindhp @nate-double-u

### CODE REVIEW

File: content/en/docs/concepts/windows/intro.md [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+The behavior of some kubelet command line options behave differently on Windows, as described below:
+
+* The `--windows-priorityclass` lets you set the scheduling priority of the kubelet process (see [CPU resource management](/docs/concepts/configuration/windows-resource-management/#resource-management-cpu))
+* The `--kubelet-reserve`, `--system-reserve` , and `--eviction-hard` flags update [NodeAllocatable](/docs/tasks/administer-cluster/reserve-compute-resources/#node-allocatable)
+* Eviction by using `--enforce-node-allocable` is not implemented
+* Eviction by using `--eviction-hard` and `--eviction-soft` are not implemented
+* A kubelet running on a Windows node does not have memory
+  restrictions. `--kubelet-reserve` and `--system-reserve` do not set limits on
+  kubelet or processes running on the host. This means kubelet or a process on the host
+  could cause memory resource starvation outside the node-allocatable and scheduler.
```
Comment (tengqm) [inline content/en/docs/concepts/windows/intro.md:158]: “This means kubelet or a process on the host could cause memory resource starvation outside the node-allocatable and scheduler.”

This sentence is difficult to parse.

File: content/en/docs/concepts/windows/intro.md [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+* The `MemoryPressure` Condition is not implemented
+* The kubelet does not take OOM eviction actions
+
+### API compatibility {#api}
+
+There are no differences in how most of the Kubernetes APIs work for Windows. The
+subtleties around what's different come down to differences in the OS and container
+runtime. In certain situations, some properties on workload resources were designed
+under the assumption that they would be implemented on Linux, and fail to run on Windows.
```
Comment (tengqm) [inline content/en/docs/concepts/windows/intro.md:167]: This is a little bit verbose? How about:

```suggestion
There are subtle differences in the way the Kubernetes APIs work for Windows due to the OS
and container runtime. Some workload properties were designed for Linux, and fail to run on Windows.
```

File: content/en/docs/concepts/services-networking/dual-stack.md [hunk 3]
```diff
@@ -239,6 +239,20 @@ If you want to enable egress traffic in order to reach off-cluster destinations
 Ensure your {{< glossary_tooltip text="CNI" term_id="cni" >}} provider supports IPv6.
 {{< /note >}}
 
+## Windows support
+
+Kubernetes on Windows does not support single-stack "IPv6-only" networking. However,
+dual-stack IPv4/IPv6 networking for pods and nodes with single-family services
+is supported.
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/concepts/services-networking/dual-stack.md:246]: How should _this_ paragraph end?
```markdown
If your cluster is dual-stack IPv6 and IPv4, Windows nodes work as expected. If you
set up single-stack IPv6-only networking for a cluster, and you add a Windows node
to that cluster,
[snakes come out of the network port / etc].
```

This limitation needs mentioning elsewhere really. People who set up single-stack clusters
may not think to consult the dual-stack page to look for Windows-specific limitations.

[the remark this answers is not in the document]
Comment (jsturtevant) [reply content/en/docs/concepts/services-networking/dual-stack.md:246]: @daschott could you help parse the intent of the paragraph?  This was a straight copy of https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#ipv6-networking.

I think the plan was to mostly move things as is in this first iteration but this one is pretty confusing.  Another thing missing here is what version of Windows is required.  WS2019 does not support dual-stack. It is available in 20H2 and WS2022.

File: content/en/docs/concepts/services-networking/windows-networking.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+* Pod → external (DNS)
+* Node → Pod
+* Pod → Node
+
+## IP address management (IPAM) {#ipam}
+
+The following IPAM options are supported on Windows:
+
+* [host-local](https://github.com/containernetworking/plugins/tree/master/plugins/ipam/host-local)
+* HNS IPAM (Inbox platform IPAM, this is a fallback when no IPAM is set)
```
[the remark this answers is not in the document]
Comment (jsturtevant) [reply content/en/docs/concepts/services-networking/windows-networking.md:93]: HNS (host network service)

**Review by jsturtevant — COMMENTED** [review]
lgtm for sig-windows, The initial plan is move these files with minimal changes to the correct content locations to make edits to them easier and make the Windows Introduction page easier to manage and comprehend.

File: content/en/docs/concepts/services-networking/windows-networking.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/services-networking/windows-networking.md:93]: I'll re-word to convey HNS is part of Windows.

File: content/en/docs/concepts/services-networking/dual-stack.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/services-networking/dual-stack.md:246]: Yes - I was hoping to move existing content then follow up with more in-depth updates. 
I'm not less familiar with networking specifics on Windows vs other parts.

I'm OK iterating here too but just think smaller PRs are easier to get merged :)

File: content/en/docs/concepts/services-networking/windows-networking.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+| Direct Server Return (DSR) | Load balancing mode where the IP address fixups and the LBNAT occurs at the container vSwitch port directly; service traffic arrives with the source IP set as the originating pod IP. | v1.20+ | Windows Server 2019 | Set the following flags in kube-proxy: `--feature-gates="WinDSR=true" --enable-dsr=true` |
+| Preserve-Destination | Skips DNAT of service traffic, thereby preserving the virtual IP of the target service in packets reaching the backend Pod. Also disables node-node forwarding. | v1.20+ | Windows Server, version 1903 (or higher) | Set `"preserve-destination": "true"` in service annotations and enable DSR in kube-proxy. |
+| IPv4/IPv6 dual-stack networking | Native IPv4-to-IPv4 in parallel with IPv6-to-IPv6 communications to, from, and within a cluster | v1.19+ | Windows Server, version 2019 | See [IPv4/IPv6 dual-stack](#ipv4ipv6-dual-stack) |
+| Client IP preservation | Ensures that source IP of incoming ingress traffic gets preserved. Also disables node-node forwarding. | v1.20+ | Windows Server, version 2019  | Set `service.spec.externalTrafficPolicy` to "Local" and enable DSR in kube-proxy |
+{{< /table >}}
```
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/services-networking/windows-networking.md:121]: @daschott / @jayunit100 are you all OK dropping the supported K8s version columns for networking functionality?

File: content/en/docs/concepts/storage/volumes.md [hunk 6]
```diff
@@ -1240,6 +1240,21 @@ for more information.
 For more information on how to develop a CSI driver, refer to the
 [kubernetes-csi documentation](https://kubernetes-csi.github.io/docs/)
 
+#### Windows CSI proxy
+
+{{< feature-state for_k8s_version="v1.22" state="stable" >}}
+
+CSI node plugins (especially those associated with persistent volumes exposed as
+either block devices or over a shared file-system) need to perform various privileged
```
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/storage/volumes.md:1248]: This was a copy/paste from intro-windows-in-kubernetes.md. I'm not sure why the dropped text was added either.

File: content/en/docs/concepts/storage/windows-storage.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+Windows has a layered filesystem driver to mount container layers and create a copy
+filesystem based on NTFS. All file paths in the container are resolved only within
+the context of that container.
+
+* With Docker, volume mounts can only target a directory in the container, and not
+  an individual file. This limitation does not exist with CRI-containerD runtime.
+* Volume mounts cannot project files or directories back to the host filesystem.
+* Read-only filesystems are not supported because write access is always required
+  for the Windows registry and SAM database. However, read-only volumes are supported.
```
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/storage/windows-storage.md:29]: Good question - let me try this out...

File: content/en/docs/concepts/services-networking/windows-networking.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/services-networking/windows-networking.md:122]: I'm dropping the supported k8s version and also fixing up the supported Windows OS versions.

File: content/en/docs/concepts/services-networking/dns-pod-service.md [hunk 8]
```diff
@@ -323,8 +324,25 @@ If the feature gate `ExpandedDNSConfig` is enabled for the kube-apiserver and
...
+  Windows treats all names with a `.` as a FQDN and skips FQDN resolution.
+- On Windows, there are multiple DNS resolvers that can be used. As these come with
+  slightly different behaviors, using the
+  [`Resolve-DNSName`](https://docs.microsoft.com/powershell/module/dnsclient/resolve-dnsname)
+  powershell cmdlet for name query resolutions is recommended.
+- On Linux, you have a DNS suffix list, which is used after resolution of a name as fully
+  qualified has failed
+  On Windows, you can only have 1 DNS suffix, which is the DNS suffix associated with that
+  pod's namespace (mydns.svc.cluster.local for example). Windows can resolve FQDNs
+  and services or names resolvable with just that suffix. For example, a pod spawned
```
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/services-networking/dns-pod-service.md:339]: I tried to make this less confusing. Let me know if it makes more sense.

File: content/en/docs/concepts/storage/windows-storage.md [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+As a result, the following storage functionality is not supported on Windows nodes:
+
+* Volume subpath mounts: only the entire volume can be mounted in a Windows container
+* Subpath volume mounting for Secrets
+* Host mount projection
+* Read-only root filesystem (mapped volumes still support `readOnly`)
+* Block device mapping
+* Memory as the storage medium (for example, `emptyDir.medium` set to `Memory`)
+* File system features like uid/gid; per-user Linux filesystem permissions
+* DefaultMode (due to UID/GID dependency)
```
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/storage/windows-storage.md:43]: I believe it is referring to https://kubernetes.io/docs/concepts/configuration/secret/#secret-files-permissions

File: content/en/docs/concepts/windows/intro.md [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+
+  A Pod is the basic building block of Kubernetes–the smallest and simplest unit in
+  the Kubernetes object model that you create or deploy. You may not deploy Windows and
+  Linux containers in the same Pod. All containers in a Pod are scheduled onto a single
+  Node where each Node represents a specific platform and architecture. The following
+  Pod capabilities, properties and events are supported with Windows containers:
+
+  * Single or multiple containers per Pod with process isolation and volume sharing
+  * Pod `status` fields
+  * Readiness and Liveness probes
```
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/windows/intro.md:80]: It looks like we only run e2e tests for liveness and readiness probes for windows (which are marked as NodeConfromance tests)

Let me run some of the e2e test cases manually to test this out.

https://github.com/kubernetes/kubernetes/blob/12990dec4057847bf6e7e8037c27baa16e58648b/test/e2e/common/node/container_probe.go

File: content/en/docs/concepts/windows/intro.md [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+  * postStart & preStop container lifecycle events
+  * ConfigMap, Secrets: as environment variables or volumes
+  * `emptyDir` volumes
+  * Named pipe host mounts
+  * Resource limits
```
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/windows/intro.md:85]: The kubelet doesn't do anything with requests - only limits. The scheduler may still take requests into consideration when assigning pods to nodes.

**Comment by marosset** [comment]
@tengqm I have incorporated all of the suggested edits. 
Since this PR is primarily moving pages around in an attempt to break up the giant intro-windows-in-kubernetes page (which is hard to edit and consume) would you be OK if the rest of the suggestions get addressed in a seperate PR?

**Comment by tengqm** [comment]
> @tengqm I have incorporated all of the suggested edits. Since this PR is primarily moving pages around in an attempt to break up the giant intro-windows-in-kubernetes page (which is hard to edit and consume) would you be OK if the rest of the suggestions get addressed in a seperate PR?

Sure. I'm okay with whichever approach to improve these pages, provided that the ball is not dropped.

**Comment by marosset** [comment]
> > @tengqm I have incorporated all of the suggested edits. Since this PR is primarily moving pages around in an attempt to break up the giant intro-windows-in-kubernetes page (which is hard to edit and consume) would you be OK if the rest of the suggestions get addressed in a seperate PR?
> 
> Sure. I'm okay with whichever approach to improve these pages, provided that the ball is not dropped.

Understood.
I will follow up with a PR to update the intro-windows page.
I will try and find someone with more networking knowledge to update that page and if not I'll try to update that too.

File: content/en/docs/concepts/storage/windows-storage.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (marosset) [reply content/en/docs/concepts/storage/windows-storage.md:29]: I'm not sure what the intention behind the existing docs are but setting volumeMounts as `readOnly: true` worked as expected for me.

I tried creating a few different types of mounts (hostPast, persistant volume based on azuredisk) and if it set `readOnly: true` I could view the contents of the volume but not modify or create content.
If i set `readOnly: false` then I could modify and create content in there. 

I can update (hopefully in a follow up PR tho :) )

**Comment by marosset** [comment]
@sftim @tengqm - do you see any issues with the structural changes in this PR?
I can open up new PRs for content changes after this merges.

**Comment by tengqm** [comment]
I'm okay to move this forward.
/approve
/label tide/merge-method-squash

File: content/en/docs/concepts/windows/user-guide.md [hunk 12]
```diff
@@ -181,29 +183,37 @@ appropriate operating system.
...
 so taints and tolerations and node selectors are still required
  to ensure that the Windows pods land onto appropriate Windows nodes.
+=======
+The scheduler does not use the value of `.spec.os.name` when assigning Pods to nodes.
+You should use normal Kubernetes mechanisms for
+[assigning pods to nodes](/docs/concepts/scheduling-eviction/assign-pod-node/)
+to ensure that the control plane for your cluster places pods onto nodes that are running the
+appropriate operating system.
+{{< /note >}}
+>>>>>>> Moving windows containers user guide out of /setup/production-environment/:content/en/docs/concepts/windows/user-guide.md
```
Comment (sftim) [inline content/en/docs/concepts/windows/user-guide.md:193]: This looks like a broken merge that needs tidying.

**Review by sftim — CHANGES_REQUESTED** [review]
/lgtm cancel

There's one place where this needs a fix.

File: content/en/docs/concepts/windows/user-guide.md (hunk 12 above)
Comment (marosset) [reply content/en/docs/concepts/windows/user-guide.md]: Fixed, I'll scan the rest of the PR again now too.

**Pull request merged**